docs/reference/filters.md

```diff
@@ -485,6 +485,21 @@ Example:
 * -> randomContent(42) -> <shunt>;
 ```
 
+## repeatContent
+
+Generate response of specified size from repeated text.
```

Comment:
Maybe a hint on memory usage would be great, because you can easily explode in memory `n=len(text) m=size: O(n*m)`

Reply:
The implementation does not allocate any buffers:
```
$ bin/skipper -inline-routes='* -> repeatContent("Hello", 100000000000) -> <shunt>' --write-timeout-server=10m
[APP]INFO[0000] Expose metrics in codahale format            
[APP]INFO[0000] support listener on :9911                    
[APP]INFO[0000] proxy listener on :9090                      
[APP]INFO[0000] TLS settings not found, defaulting to HTTP   
[APP]INFO[0000] route settings, reset, route: : * -> repeatContent("Hello", 100000000000) -> <shunt> 
[APP]INFO[0000] route settings received                      
[APP]INFO[0000] route settings applied                       
127.0.0.1 - - [24/Nov/2020:00:31:30 +0100] "GET / HTTP/1.1" 200 100000000000 "-" "curl/7.58.0" 153980 localhost:9090 - -
```
```
$ curl localhost:9090 -o /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 93.1G    0 93.1G    0     0   619M      0 --:--:--  0:02:33 --:--:--  615M
```

Reply:
I think some background is missing on why I propose this filter.

While working on https://github.com/zalando/skipper/pull/1567 I have discovered there is no suitable way to generate response of a given size.

I have tried `randomContent` filter but it had low performance which was later improved by #1570 but is still subpar due to relatively slow random generator. 

Another option is to use fixed-sized files via `static` filter but it [allocates buffers](https://github.com/zalando/skipper/pull/1567#issuecomment-716702302) to copy file content and is not convenient as one has to prepare files. 

This implementation does not allocate any buffers but fills up a buffer provided by proxy's `copyStream` with repeated pattern using a fast copy-from-itself trick (like [bytes.Repeat](https://golang.org/pkg/bytes/#Repeat) does) - e.g. it copies `hello` into the buffer, then copies `hello` from the buffer into itself resulting in `hellohello` then again resulting in `hellohellohellohello` etc. So e.g. `repeatContent("x", 10000)` fills up default proxy's buffer of 8192 size just in log2(8192)=13 iterations (contrary to naive for-loop implementation which would require 8192 iterations).

I think this implementation is the fastest way to serve content of a given size in skipper. Serving file could be faster because of [`io.ReaderFrom`](https://golang.org/src/net/http/server.go#L581) which utilized `sendfile` if we would use `io.CopyBuffer` or support `io.ReaderFrom` in proxy's `copyStream` instead of hand-rolled loop.

Reply:
Here are some numbers for serving 8k response:
```
$ # randomContent
$ bin/skipper -inline-routes='* -> randomContent(8192) -> <shunt>' -access-log-disabled
$ echo "GET http://localhost:9090/" | vegeta attack -duration=1m -rate=0 --max-workers=500 | vegeta report 
Requests      [total, rate, throughput]         358698, 5978.31, 5970.25
Duration      [total, attack, wait]             1m0s, 1m0s, 80.97ms
Latencies     [min, mean, 50, 90, 95, 99, max]  217.34µs, 83.591ms, 88.377ms, 92.926ms, 94.426ms, 99.129ms, 276.118ms
Bytes In      [total, mean]                     2938454016, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:358698  
Error Set:
```
```
$ # static
$ truncate -s 8k 8k.bin
$ bin/skipper -inline-routes='* -> static("/", "./") -> <shunt>' -access-log-disabled
$ echo "GET http://localhost:9090/8k.bin" | vegeta attack -duration=1m -rate=0 --max-workers=500 | vegeta report 
Requests      [total, rate, throughput]         1010975, 16849.50, 16847.79
Duration      [total, attack, wait]             1m0s, 1m0s, 6.089ms
Latencies     [min, mean, 50, 90, 95, 99, max]  114.431µs, 12.959ms, 9.043ms, 30.901ms, 38.343ms, 54.711ms, 122.876ms
Bytes In      [total, mean]                     8281907200, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:1010975  
Error Set:
```
```
$ repeatContent
$ bin/skipper -inline-routes='* -> repeatContent("x", 8192) -> <shunt>' -access-log-disabled
$ echo "GET http://localhost:9090/" | vegeta attack -duration=1m -rate=0 --max-workers=500 | vegeta report 
Requests      [total, rate, throughput]         1155795, 19262.71, 19259.43
Duration      [total, attack, wait]             1m0s, 1m0s, 10.216ms
Latencies     [min, mean, 50, 90, 95, 99, max]  100.193µs, 11.104ms, 7.197ms, 27.356ms, 33.557ms, 45.277ms, 134.422ms
Bytes In      [total, mean]                     9468272640, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:1155795  
Error Set:
```

Reply:
The memory size in skipper would be the size of the input string and the buffers would increase to number of unwritten bytes to the socket.
Ok great to see, had to check again that you work on the one byte slice and read in the wrapper io.LimitReader